lib/Optimizer/Transforms/DelayMeasurements.cpp

```diff
@@ -0,0 +1,101 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+
+#define DEBUG_TYPE "delay-measurements"
+
+using namespace mlir;
+
+namespace cudaq::opt {
+
+/// Delay Measurements
+///
+/// This pass delays measurements as long as possible. This reordering of
+/// instructions is useful for Base Profile programs where the user measures 1
+/// qubit early in the program and then performs quantum operations on other
+/// qubits later in the program.
+struct DelayMeasurementsPass
+    : public cudaq::opt::DelayMeasurementsBase<DelayMeasurementsPass> {
+  // explicit DelayMeasurementsPass() = default;
+
+  void runOnOperation() override {
+
+    auto func = getOperation();
+    auto &blocks = func.getBlocks();
+
+    if (blocks.size() == 0)
+      return;
+
+    if (blocks.size() > 1) {
+      func.emitError("DelayMeasurementsPass cannot handle multiple blocks. Do "
+                     "you have if statements in a Base Profile QIR program?");
+      signalPassFailure();
+      return;
+    }
+
+    moveMeasurementsToEnd(*blocks.begin());
+  }
+
+  /// @brief Add `op` and all of its users into `opsToMoveToEnd`
+  void
+  addOpAndUsersToList(Operation &op,
+                      llvm::SmallVector<mlir::Operation *> &opsToMoveToEnd) {
+    opsToMoveToEnd.push_back(&op);
+    for (auto user : op.getUsers())
+      addOpAndUsersToList(*user, opsToMoveToEnd);
+  }
+
+  /// @brief The Base Profile requires that irreversible operations (i.e.
+  /// measurements) come after reversible operations. This function enforces
+  /// that.
+  /// @param mainBlock block to process
+  void moveMeasurementsToEnd(mlir::Block &mainBlock) {
+    llvm::SmallVector<mlir::Operation *> opsToMoveToEnd;
+
+    // Keep track of which qubits have been measured as we're walking through
+    // the block
+    llvm::DenseSet<mlir::Value> measuredQubits;
+
+    // Step 1: Identify operations to move. Add to opsToMoveToEnd.
+    for (auto &op : mainBlock) {
+      if (op.hasTrait<QuantumMeasure>()) {
+        // Save the fact that we're measuring these qubits
+        for (auto operand : op.getOperands())
+          measuredQubits.insert(operand);
+      }
+
+      if (op.hasTrait<QuantumMeasure>() || isa<mlir::func::ReturnOp>(op) ||
+          isa<quake::DeallocOp>(op)) {
```

Comment:
So this pass isn't going to work with `!quake.wire` values, it seems.

It seems like it would be clean to implement something like this in the value semantics. A measurement and all the code that has data flow originating from that measurement can be sunk past any operations that are data flow independent in the block.

Reply:
Some do, but it's an obvious limitation on those passes.

Generally, we need to stop baking in this limitation and make passes that "work right" (TM).

Reply:
Actually, in re-review it this morning, I think the recursive calls in `addOpAndUsersToList()` already handle this. I added a test in e81199ef to prove to myself that it still works with value semantics. Please let me know if you think there is something that you specifically spotted in the code that makes you think it won't work as written because I didn't have to make any changes to make the new test work.

Reply:
What raised a blip on my radar was the quake::DeallocOp. Those ops are normally erased in the quantum value form (replaced with quake::SinkOp). I suppose it is possible that algorithmically, since SinkOps will be users of the measurements (perhaps), it'll be ok as is.